### PR TITLE
712 - `shinytest2` for `tm_variable_browser`

### PR DESCRIPTION
**Pull request by averissimo**
# Pull Request

Part of  #712 

#### Changes description

- prefixes `ggplot2` functions calls
- "content is displayed correctly" test
  - Ensures that all datasets are present in left table
  - Ensures that table below the plot has correct first column title ("Statistic" for numeric variables and "levels" for categorical)
- "main output interactivity doesn't show errors" test
  - Tests all buttons and input sliders

#### Notes:

- Click function on javascript elements with callbacks doesn't trigger them
  - Clicking on variable table row does not select them

### CODE REVIEW

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        grepl("fa-chart-bar", app_driver$get_attr(categorical_selector, "class")),
+        arr.ind = TRUE
+      ),
+      Inf
+    ),
+    na.rm = TRUE
+  )
+
+  if (!is.infinite(categorical_index)) {
+    app_driver$set_module_input(
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:54]:
```suggestion
    app_driver$set_active_module_input(
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      sprintf("variable_browser_%s_rows_selected", current_var),
+      categorical_index,
+      allow_no_input_binding_ = TRUE
+    )
+    app_driver$set_module_input(
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:59]:
```suggestion
    app_driver$set_active_module_input(
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  ) ## Does not exist initially
+
+  app_driver$click(selector = selector_ns("remove_outliers"))
+  app_driver$expect_no_shiny_error()
+  testthat::expect_length(
+    app_driver$active_module_element_text("outlier_definition_slider"),
+    1
+  ) ## Added to UI
+
+  app_driver$set_module_input("outlier_definition_slider", 2)
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:103]:
```suggestion
  app_driver$set_active_module_input("outlier_definition_slider", 2)
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  app_driver$expect_no_shiny_error()
+
+  app_driver$click(selector = selector_ns("remove_outliers"))
+  app_driver$expect_no_shiny_error()
+
+  # Test changing plot settings
+  testthat::expect_false(app_driver$is_visible(selector_ns("ggplot_theme-selectized")))
+
+  app_driver$set_module_input("ggplot_theme", "bw")
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:112]:
```suggestion
  app_driver$set_active_module_input("ggplot_theme", "bw")
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("ggplot_theme", "light")
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:114]:
```suggestion
  app_driver$set_active_module_input("ggplot_theme", "light")
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("ggplot_theme", "dark")
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:116]:
```suggestion
  app_driver$set_active_module_input("ggplot_theme", "dark")
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  app_driver$expect_no_validation_error()
+
+  app_driver$set_module_input("font_size", "8")
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:119]:
```suggestion
  app_driver$set_active_module_input("font_size", "8")
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("font_size", "20")
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:121]:
```suggestion
  app_driver$set_active_module_input("font_size", "20")
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("font_size", "15")
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:123]:
```suggestion
  app_driver$set_active_module_input("font_size", "15")
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+  app_driver$expect_no_validation_error()
+
+  app_driver$set_module_input("label_rotation", "25")
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:126]:
```suggestion
  app_driver$set_active_module_input("label_rotation", "25")
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("label_rotation", "0")
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:128]:
```suggestion
  app_driver$set_active_module_input("label_rotation", "0")
```

File: tests/testthat/test-shinytest2-tm_variable_browser.R [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+  app_driver$expect_no_validation_error()
+  app_driver$set_module_input("label_rotation", "90")
```
Comment (m7pr) [inline tests/testthat/test-shinytest2-tm_variable_browser.R:130]:
```suggestion
  app_driver$set_active_module_input("label_rotation", "90")
```

**Review by m7pr — CHANGES_REQUESTED** [review]
rename `set_module_input` to `set_active_module_input`

**Review by kartikeyakirar — CHANGES_REQUESTED** [review]
@averissimo test passes with warning. 
image

Apart from that few additional request:

1: We don't need to call app_driver$expect_no_shiny_error() after every initialization.
2: Could you add tests to check the visibility of plots using app_driver$is_visible()?
2. is is possible to add test for checkbox (teat variable as factor)
image

**Comment by averissimo** [comment]
Thanks for the comments @kartikeyakirar.

I implemented them on the two new commits and split some of the tests that were too long / big scope.

> 1: We don't need to call app_driver$expect_no_shiny_error() after every initialization.

Does it seem fine as it is, or did you also mean to remove some of them after changing the inputs?

**Comment by kartikeyakirar** [comment]
> Does it seem fine as it is, or did you also mean to remove some of them after changing the inputs?

I'm fine with either option, but since this additional test has already been checked and Marcin has removed it from another PR, let's keep it consistent.

**Review by kartikeyakirar — APPROVED** [review]
Great Work

**Pull request merged**